### PR TITLE
Problem: assert used for ZAP error handling aborts process.

### PR DESCRIPTION
**Pull request by evoskuil**
See: https://github.com/zeromq/libzmq/issues/2476

I've followed the general existing pattern whereby when a failure results when sending a message the `zmq_msg_t` object is not closed. This looks like a leak to me but it's a more general question than I wanted to resolve in this pull request. Worst case we are temporarily trading a crash for a leak in the case of an error that apparently isn't very common.

### CODE REVIEW

**Comment by bluca** [comment]
Why is there a massive diff in src/curve_server.cpp ?

**Comment by evoskuil** [comment]
Yes, it results from the interaction with git automatic CRLF/LF conversion. When a file has a mixture of LF and CRLF characters (which this one does) git does not convert it to CRLF. But then the IDE does convert it, git doesn't then reset it to all LF (because it wasn't initially converted due to being mixed). :/

**Comment by bluca** [comment]
Any chance you could fix it so that the commit doesn't change every line please? Git blame is a really important tool

File: src/curve_server.cpp [hunk 1]
```diff
@@ -310,12 +310,12 @@ int zmq::curve_server_t::process_hello (msg_t *msg_)
                               sizeof hello_box,
                               hello_nonce, cn_client, secret_key);
     if (rc != 0) {
-        //  Temporary support for security debugging
-        puts("CURVE I: cannot open client HELLO -- wrong server key?");
-        errno = EPROTO;
+        //  Temporary support for security debugging
+        puts ("CURVE I: cannot open client HELLO -- wrong server key?");
+        errno = EPROTO;
```
Comment (evoskuil) [inline src/curve_server.cpp]: This section had \r\n line endings. Converted to \n (and fixed style).

**Comment by evoskuil** [comment]
All set, squashed commits.

**Comment by bluca** [comment]
Great, thanks!

**Comment by bluca** [comment]
Regarding closing a message on failure to send, usually an error is returned to the caller who then has to take care of it, right? (On the phone so not easy to check around)

**Comment by evoskuil** [comment]
Regarding message closure on failure, I created https://github.com/zeromq/libzmq/issues/2481

**Pull request merged**